abalone.hs

```diff
@@ -1,190 +1,205 @@
-module Abalone (
-	Game(Game), Board(Board), Player(White, Black), Position,
-	gameOver, winner, numPieces, futures, isValid,
-	Direction, Segment, segments, adjacent, start
-	) where
-
+module Abalone
+  ( Game(Game)
+  , Board(Board)
+  , Player(White, Black)
+  , Position
+  , gameOver
+  , winner
+  , numPieces
+  , futures
+  , isValid
+  , Direction
+  , Segment
+  , segments
+  , (|>)
+  , start
+  ) where
+
+import Data.Ord
+import Data.List
+import Data.Set (Set)
 import qualified Data.Set as Set
 
-data Game = Game {
-	board 		   :: Board,
-	nextPlayer 	   :: Player,
-	movesRemaining :: Int,
-	marblesPerMove :: Int
-} deriving (Eq, Show, Read)
+import Control.Applicative
+import Control.Monad
+
+data Game = Game { board          :: Board
+                 , nextPlayer     :: Player
+                 , movesRemaining :: Int
+                 , marblesPerMove :: Int
+                 } deriving (Eq, Show, Read)
 
-data Board = Board {
-	whitePositions :: Set.Set Position,
-	blackPositions :: Set.Set Position,
-	boardRadius    :: Int
-} deriving (Eq, Show, Read)
+data Board = Board { whitePositions :: Set Position
+                   , blackPositions :: Set Position
+                   , boardRadius    :: Int
+                   } deriving (Eq, Show, Read)
 
-getPieces :: Board -> Player -> Set.Set Position
-getPieces b p = (if p == White then whitePositions else blackPositions) b
+getPieces :: Board -> Player -> Set Position
+getPieces b White = whitePositions b
+getPieces b Black = blackPositions b
 
-start :: Game 
+start :: Game
 start = Game standardBoard White 1000 3
 
 standardBoard :: Board
-standardBoard = Board (Set.fromList whitePos) (Set.fromList blackPos) 5 where
-	whitePos = [(-x, 4) | x <- [0..4]] ++ [(-x, 3) | x <- [-1..4]] ++ [(-2, 2), (-1, 2), (0, 2)]
-	blackPos = map (\(q, r) -> (-q, -r)) whitePos
+standardBoard = Board whitePos blackPos 5
+ where
+  whitePos = Set.fromList $ [(-4,0,4),(-4,1,3),(-2,0,2)] >>= \(q,q',r) -> map (flip (,) r) [q..q']
+  blackPos = Set.map (\(q, r) -> (-q, -r)) whitePos
 
 -- Player & Related Functions--
-data Player = White | Black deriving (Eq, Show, Read, Ord, Bounded, Enum)
+data Player = White | Black
+  deriving (Eq, Show, Read, Ord, Bounded, Enum)
+
 next :: Player -> Player
-next p = case p of
-	White -> Black
-	Black -> White
+next White = Black
+next Black = White
 
 -- Position / Grid Functions --
 type Position = (Int, Int)
+
 dist2 :: Position -> Position -> Int -- distance * 2 (to avoid fractional types)
-dist2 (q1,r1) (q2,r2) = abs(q1 - q2) + abs(r1 - r2) + abs(q1 + r1 - q2 - r2)
+dist2 (q1,r1) (q2,r2) = abs (q1 - q2) + abs (r1 - r2) + abs (q1 + r1 - q2 - r2)
 
-data Direction = TopRight | MidRight | BotRight 
-			   | BotLeft  | MidLeft  | TopLeft 
-			   deriving (Eq, Show, Read, Ord, Bounded, Enum)
+data Direction = TopLeft | TopRight
+               | MidLeft | MidRight
+               | BotLeft | BotRight
+  deriving (Eq, Show, Read, Ord, Bounded, Enum)
```

Comment:
the previous order was a clockwise enumeration... not that it really mattered. maybe i should remove deriving (Ord) since I'm not sure the ordering here is at all significant.